### PR TITLE
Emacs navigation key-bindings support

### PR DESCRIPTION
**Pull request by janmejay**
This change adds emacs-style navigation keymap support (similar to existing Vi keymap support).

### CODE REVIEW

File: src/common/common.cpp [hunk 1]
```diff
@@ -498,6 +498,70 @@ bool handleViKey(QKeyEvent *event, QObject *eventReceiver)
...
+      break;
+    }
+    return false;
+  case Qt::Key_Comma:
+    if ((mods & Qt::ControlModifier) && (mods & Qt::ShiftModifier)) {
+      key = Qt::Key_Home;
+      mods = mods & ~(Qt::ShiftModifier || Qt::ControlModifier);
+    } else {
+      return false;
+    }
```
Comment (hluk) [inline src/common/common.cpp:538]: Missing break?

File: src/common/common.cpp [hunk 2]
```diff
@@ -498,6 +498,70 @@ bool handleViKey(QKeyEvent *event, QObject *eventReceiver)
...
+  case Qt::Key_Stop:
```
Comment (hluk) [inline src/common/common.cpp:539]: What key is this? Looks like special key on some keyboards to stop loading page (as opposed to `Qt::Key_Refresh`).

File: src/common/common.cpp [hunk 3]
```diff
@@ -498,6 +498,70 @@ bool handleViKey(QKeyEvent *event, QObject *eventReceiver)
...
+    if (mods & Qt::ControlModifier) {
+      key = Qt::Key_Up;
+      mods = mods & ~Qt::ControlModifier;
+      break;
+    }
+    return false;
+  case Qt::Key_Comma:
+    if ((mods & Qt::ControlModifier) && (mods & Qt::ShiftModifier)) {
+      key = Qt::Key_Home;
+      mods = mods & ~(Qt::ShiftModifier || Qt::ControlModifier);
```
Comment (hluk) [inline src/common/common.cpp:535]: I'm using custom CMake flag for pedantic compiler warnings (`cmake -DPEDANTIC=ON...`) and this gives me following error.

```
src/common/common.cpp:535: error: enum constant in boolean context [-Werror=int-in-bool-context]
       mods = mods & ~(Qt::ShiftModifier || Qt::ControlModifier);
                                                ^~~~~~~~~~~~~~~
```

```suggestion
      mods.setFlag(Qt::ShiftModifier, false);
      mods.setFlag(Qt::ControlModifier, false);
```

File: src/common/common.cpp [hunk 4]
```diff
@@ -498,6 +498,70 @@ bool handleViKey(QKeyEvent *event, QObject *eventReceiver)
...
+    } else {
+      return false;
+    }
+  case Qt::Key_Stop:
+    if ((mods & Qt::ControlModifier) && (mods & Qt::ShiftModifier)) {
+      key = Qt::Key_End;
+      mods = mods & ~(Qt::ShiftModifier || Qt::ControlModifier);
```
Comment (hluk) [inline src/common/common.cpp:542]: Same as above:

```suggestion
      mods.setFlag(Qt::ShiftModifier, false);
      mods.setFlag(Qt::ControlModifier, false);
```

File: src/gui/traymenu.cpp [hunk 5]
```diff
@@ -84,8 +85,15 @@ void TrayMenu::addClipboardItemAction(const QModelIndex &index, bool showImages,
     resetSeparators();
 
     // Show search text at top of the menu.
-    if ( m_clipboardItemActionCount == 0 && m_searchText.isEmpty() )
-        setSearchMenuItem( m_viMode ? tr("Press '/' to search") : tr("Type to search") );
+    if ( m_clipboardItemActionCount == 0 && m_searchText.isEmpty() ) {
+      if (m_viMode) {
+        setSearchMenuItem( tr("Press '/' to search") );
+      } else if (m_emacsMode) {
+        setSearchMenuItem( tr("Press 'Ctrl + s' to search") );
```
Comment (hluk) [inline src/gui/traymenu.cpp:92]: The shortcut doesn't seem to work for me in the tray menu (`copyq menu`).

File: src/gui/traymenu.cpp [hunk 6]
```diff
@@ -169,9 +177,20 @@ void TrayMenu::clearAllActions()
...
+    }
     m_viMode = enabled;
 }
 
+void TrayMenu::setEmacsModeEnabled(bool enabled)
+{
+    if (enabled) {
+        m_viMode = false;
+    }
+    m_emacsMode = enabled;
```
Comment (hluk) [inline src/gui/traymenu.cpp:191]: Something to consider is using an enum to store the current navigation style. Also might be good to change the two check boxes in preferences into a combo-box or radio button group with three options.

**Comment by scottkosty** [comment]
@janmejay Ping. I just wanted to make sure you saw the comments.

**Pull request closed without merging**